### PR TITLE
Return (bool, error) in Authorizer.Authorize()

### LINKED ISSUE #27974: Authorizer.Authorize() should return (bool, error)
The [Authorizer interface](https://github.com/kubernetes/kubernetes/blob/master/pkg/auth/authorizer/interfaces.go#L76) only returns an error today. This means that clients tend to treat any error as a "No" and return forbidden (e.g. our main [apiserver handler](https://github.com/kubernetes/kubernetes/blob/master/pkg/apiserver/handlers.go#L484)).

Our authorizers have become more complicated (and have upstream dependencies in the case of the webhook). If the authorizer cannot make a decision, the apiserver should return a retryable error (probably a 503 or 504) rather than a 403 that the client will most likely not retry.

### PR DESCRIPTION
**Pull request by nhlfr**
Before this change, Authorize() method was just returning an error, regardless of whether the user is unauthorized or whether there is some other unrelated error. Returning boolean with information about user authorization and error (which should be unrelated to the authorization) separately will make it easier to debug.

Fixes #27974

### CODE REVIEW

File: pkg/apis/rbac/validation/rulevalidation.go [hunk 1]
```diff
@@ -40,7 +39,7 @@ type AuthorizationRuleResolver interface {
 }
 
 // ConfirmNoEscalation determines if the roles for a given user in a given namespace encompass the provided role.
-func ConfirmNoEscalation(ctx api.Context, ruleResolver AuthorizationRuleResolver, rules []rbac.PolicyRule) error {
+func ConfirmNoEscalation(ctx api.Context, ruleResolver AuthorizationRuleResolver, rules []rbac.PolicyRule) bool {
```
Comment (liggitt) [inline pkg/apis/rbac/validation/rulevalidation.go]: how is this related to the authorizer interface change? I'd rather preserve the information with the error return

File: pkg/apiserver/handlers.go [hunk 2]
```diff
@@ -481,12 +485,14 @@ func WithImpersonation(handler http.Handler, requestContextMapper api.RequestCon
...
-		err := a.Authorize(getAttribs.GetAttribs(req))
-		if err == nil {
-			handler.ServeHTTP(w, req)
-			return
+		authorized, err := a.Authorize(getAttribs.GetAttribs(req))
+		if !authorized {
+			forbidden(w, req)
 		}
-		forbidden(w, req)
+		if err != nil {
```
Comment (liggitt) [inline pkg/apiserver/handlers.go]: error check should come first

File: pkg/apiserver/handlers.go [hunk 3]
```diff
@@ -481,12 +485,14 @@ func WithImpersonation(handler http.Handler, requestContextMapper api.RequestCon
...
+			internalError(w, req)
```
Comment (liggitt) [inline pkg/apiserver/handlers.go]: need to `return` early

File: pkg/apiserver/handlers.go [hunk 4]
```diff
@@ -481,12 +485,14 @@ func WithImpersonation(handler http.Handler, requestContextMapper api.RequestCon
 // WithAuthorizationCheck passes all authorized requests on to handler, and returns a forbidden error otherwise.
 func WithAuthorizationCheck(handler http.Handler, getAttribs RequestAttributeGetter, a authorizer.Authorizer) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-		err := a.Authorize(getAttribs.GetAttribs(req))
-		if err == nil {
-			handler.ServeHTTP(w, req)
-			return
+		authorized, err := a.Authorize(getAttribs.GetAttribs(req))
+		if !authorized {
+			forbidden(w, req)
```
Comment (liggitt) [inline pkg/apiserver/handlers.go]: need to `return` early

File: pkg/auth/authorizer/abac/abac.go [hunk 5]
```diff
@@ -215,13 +215,13 @@ func resourceMatches(p api.Policy, a authorizer.Attributes) bool {
...
-func (pl policyList) Authorize(a authorizer.Attributes) error {
+func (pl policyList) Authorize(a authorizer.Attributes) (bool, error) {
 	for _, p := range pl {
 		if matches(*p, a) {
-			return nil
+			return true, nil
 		}
 	}
-	return errors.New("No policy matched.")
+	return false, errors.New("No policy matched.")
```
Comment (liggitt) [inline pkg/auth/authorizer/abac/abac.go]: this shouldn't be considered an error, just a normal denial

Comment (liggitt) [inline pkg/auth/authorizer/abac/abac.go]: this is actually a good example of the `reason` field @deads2k referred to... the "No policy matched" information is useful, but not an error

File: pkg/registry/clusterrole/policybased/storage.go [hunk 6]
```diff
@@ -50,8 +52,8 @@ func (s *Storage) Create(ctx api.Context, obj runtime.Object) (runtime.Object, e
 
 	clusterRole := obj.(*rbac.ClusterRole)
 	rules := clusterRole.Rules
-	if err := validation.ConfirmNoEscalation(ctx, s.ruleResolver, rules); err != nil {
-		return nil, errors.NewForbidden(groupResource, clusterRole.Name, err)
+	if authorized := validation.ConfirmNoEscalation(ctx, s.ruleResolver, rules); !authorized {
+		return nil, errors.NewForbidden(groupResource, clusterRole.Name, fmt.Errorf(""))
```
Comment (liggitt) [inline pkg/registry/clusterrole/policybased/storage.go]: revert this change

File: pkg/registry/clusterrole/policybased/storage.go [hunk 7]
```diff
@@ -67,8 +69,8 @@ func (s *Storage) Update(ctx api.Context, name string, obj rest.UpdatedObjectInf
 		clusterRole := obj.(*rbac.ClusterRole)
 
 		rules := clusterRole.Rules
-		if err := validation.ConfirmNoEscalation(ctx, s.ruleResolver, rules); err != nil {
-			return nil, errors.NewForbidden(groupResource, clusterRole.Name, err)
+		if authorized := validation.ConfirmNoEscalation(ctx, s.ruleResolver, rules); !authorized {
+			return nil, errors.NewForbidden(groupResource, clusterRole.Name, fmt.Errorf(""))
```
Comment (liggitt) [inline pkg/registry/clusterrole/policybased/storage.go]: and this

File: pkg/registry/clusterrolebinding/policybased/storage.go [hunk 8]
```diff
@@ -53,7 +53,7 @@ func (s *Storage) Create(ctx api.Context, obj runtime.Object) (runtime.Object, e
 	if err != nil {
 		return nil, err
 	}
-	if err := validation.ConfirmNoEscalation(ctx, s.ruleResolver, rules); err != nil {
+	if authorized := validation.ConfirmNoEscalation(ctx, s.ruleResolver, rules); !authorized {
```
Comment (liggitt) [inline pkg/registry/clusterrolebinding/policybased/storage.go]: revert

File: pkg/registry/role/policybased/storage.go [hunk 9]
```diff
@@ -50,8 +52,8 @@ func (s *Storage) Create(ctx api.Context, obj runtime.Object) (runtime.Object, e
 
 	role := obj.(*rbac.Role)
 	rules := role.Rules
-	if err := validation.ConfirmNoEscalation(ctx, s.ruleResolver, rules); err != nil {
-		return nil, errors.NewForbidden(groupResource, role.Name, err)
+	if authorized := validation.ConfirmNoEscalation(ctx, s.ruleResolver, rules); !authorized {
```
Comment (liggitt) [inline pkg/registry/role/policybased/storage.go]: revert

File: pkg/registry/rolebinding/policybased/storage.go [hunk 10]
```diff
@@ -53,7 +53,7 @@ func (s *Storage) Create(ctx api.Context, obj runtime.Object) (runtime.Object, e
 	if err != nil {
 		return nil, err
 	}
-	if err := validation.ConfirmNoEscalation(ctx, s.ruleResolver, rules); err != nil {
+	if authorized := validation.ConfirmNoEscalation(ctx, s.ruleResolver, rules); !authorized {
```
Comment (liggitt) [inline pkg/registry/rolebinding/policybased/storage.go]: revert

File: plugin/pkg/auth/authorizer/rbac/rbac.go [hunk 11]
```diff
@@ -62,7 +62,7 @@ func (r *RBACAuthorizer) Authorize(attr authorizer.Attributes) error {
 		}
 	}
 
-	return validation.ConfirmNoEscalation(ctx, r.authorizationRuleResolver, []rbac.PolicyRule{requestedRule})
+	return validation.ConfirmNoEscalation(ctx, r.authorizationRuleResolver, []rbac.PolicyRule{requestedRule}), nil
```
Comment (liggitt) [inline plugin/pkg/auth/authorizer/rbac/rbac.go]: for now, handle an error returned from ConfirmNoEscalation as a denial. @ericchiang, this is probably not how we want to do authorization checks for every request... escalation checks are a pretty heavyweight way to check specific attributes against a set of rules

File: plugin/pkg/auth/authorizer/webhook/webhook.go [hunk 12]
```diff
@@ -180,10 +186,10 @@ func (w *WebhookAuthorizer) Authorize(attr authorizer.Attributes) (err error) {
 		}
 	}
 	if r.Status.Allowed {
-		return nil
+		return true, nil
 	}
 	if r.Status.Reason != "" {
-		return errors.New(r.Status.Reason)
+		return false, errors.New(r.Status.Reason)
```
Comment (liggitt) [inline plugin/pkg/auth/authorizer/webhook/webhook.go]: I think we want a return type of `bool, string, error` to allow preserving this info

**Comment by liggitt** [comment]
While we're rippling an interface change, I'd like the context added to the signature for request tracing purposes. Combined with an optional reason for denials (which the current code clearly wants in a couple places), I think we end up with:

```
Authorize(ctx api.Context, attrs authorizer.Attributes) (authorized bool, reason string, err error)
```

I know @deads2k is happy with that, since we speak as one. @erictune, @cjcullen, that seem reasonable to you?

File: plugin/pkg/auth/authorizer/rbac/rbac.go (hunk 11 above)
Comment (ericchiang) [inline plugin/pkg/auth/authorizer/rbac/rbac.go]: This seemed like the correct way to answer the question "does this context have access to this resource" (just pass a single rule rather than multiple). Don't we have to ask that question in some capacity, even if it's not through the rule resolver?

Comment (liggitt) [inline plugin/pkg/auth/authorizer/rbac/rbac.go]: @ericchiang synthesizing a rule from attributes and doing an escalation check on that rule seems both expensive (since the escalation check is geared toward many-to-many permission checks) and error prone (looking at this more, the non-resource-url rule is not built correctly since it leaves out verb). to avoid hijacking this PR, we can split this discussion into a separate issue

Comment (liggitt) [inline plugin/pkg/auth/authorizer/rbac/rbac.go]: opened https://github.com/kubernetes/kubernetes/issues/28291 for the non-resource-url issue

Comment (ericchiang) [inline plugin/pkg/auth/authorizer/rbac/rbac.go]: :+1: I've opened https://github.com/kubernetes/kubernetes/issues/28292

**Comment by nhlfr** [comment]
@liggitt Done.

I will write and fix integration tests soon.

File: pkg/apiserver/authz.go [hunk 13]
```diff
@@ -55,24 +55,38 @@ func NewAlwaysAllowAuthorizer() authorizer.Authorizer {
...
+}
+
+func NewAlwaysFailAuthorizer() authorizer.Authorizer {
+	return new(alwaysFailAuthorizer)
+}
+
 const (
 	ModeAlwaysAllow string = "AlwaysAllow"
 	ModeAlwaysDeny  string = "AlwaysDeny"
+	ModeAlwaysFail  string = "AlwaysFail"
```
Comment (liggitt) [inline pkg/apiserver/authz.go]: This is useful for unit tests but I'm not sure we'd want to ever surface it as an actual startup option

File: pkg/apiserver/handlers.go [hunk 14]
```diff
@@ -447,11 +447,15 @@ func WithImpersonation(handler http.Handler, requestContextMapper api.RequestCon
 			actingAsAttributes.Name = name
 		}
 
-		err := a.Authorize(actingAsAttributes)
-		if err != nil {
+		authorized, _, err := a.Authorize(actingAsAttributes)
+		if !authorized {
```
Comment (liggitt) [inline pkg/apiserver/handlers.go]: Handle `err != nil` case first

File: pkg/apiserver/handlers.go [hunk 15]
```diff
@@ -447,11 +447,15 @@ func WithImpersonation(handler http.Handler, requestContextMapper api.RequestCon
...
 			forbidden(w, req)
```
Comment (liggitt) [inline pkg/apiserver/handlers.go]: Logging the forbidden reason here seems useful

File: pkg/apiserver/handlers.go [hunk 16]
```diff
@@ -481,12 +485,16 @@ func WithImpersonation(handler http.Handler, requestContextMapper api.RequestCon
 // WithAuthorizationCheck passes all authorized requests on to handler, and returns a forbidden error otherwise.
 func WithAuthorizationCheck(handler http.Handler, getAttribs RequestAttributeGetter, a authorizer.Authorizer) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-		err := a.Authorize(getAttribs.GetAttribs(req))
-		if err == nil {
-			handler.ServeHTTP(w, req)
+		authorized, _, err := a.Authorize(getAttribs.GetAttribs(req))
+		if !authorized {
```
Comment (liggitt) [inline pkg/apiserver/handlers.go]: Err case first, log the forbidden reason if !authorized

File: plugin/pkg/auth/authorizer/webhook/webhook_test.go [hunk 17]
```diff
@@ -508,33 +524,77 @@ func TestWebhookCache(t *testing.T) {
 	attr := authorizer.AttributesRecord{User: &user.DefaultInfo{Name: "alice"}}
 	serv.allow = true
 	serv.statusCode = 500
-	if err := wh.Authorize(attr); err == nil {
+	authorized, _, err := wh.Authorize(attr)
```
Comment (ericchiang) [inline plugin/pkg/auth/authorizer/webhook/webhook_test.go]: When possible prefer table driven tests[0] instead of rewriting these over and over again. 

[0] https://github.com/golang/go/wiki/TableDrivenTests

File: pkg/auth/authorizer/union/union.go [hunk 18]
```diff
@@ -30,16 +30,25 @@ func New(authorizationHandlers ...authorizer.Authorizer) authorizer.Authorizer {
 }
 
 // Authorizes against a chain of authorizer.Authorizer objects and returns nil if successful and returns error if unsuccessful
-func (authzHandler unionAuthzHandler) Authorize(a authorizer.Attributes) error {
-	var errlist []error
+func (authzHandler unionAuthzHandler) Authorize(a authorizer.Attributes) (bool, string, error) {
+	var (
+		errlist []error
+		reason  string
```
Comment (liggitt) [inline pkg/auth/authorizer/union/union.go]: reasonlist []string

File: pkg/auth/authorizer/union/union.go [hunk 19]
```diff
@@ -30,16 +30,25 @@ func New(authorizationHandlers ...authorizer.Authorizer) authorizer.Authorizer {
...
+	)
 	for _, currAuthzHandler := range authzHandler {
-		err := currAuthzHandler.Authorize(a)
+		authorized, currentReason, err := currAuthzHandler.Authorize(a)
+		if currentReason != "" {
```
Comment (liggitt) [inline pkg/auth/authorizer/union/union.go]: keep a list of reasons, append to it in unauthorized cases (and also return the reason in authorized cases)
```
authorized, reason, err := currAuthzHandler.Authorize(a)
if err != nil {
  errlist = append(errlist, err)
  continue
}
if !authorized {
  if reason != "" {
    reasonlist = append(reasonlist, reason)
  }
  continue
}
return true, reason, nil
```

File: pkg/auth/authorizer/union/union.go [hunk 20]
```diff
@@ -30,16 +30,25 @@ func New(authorizationHandlers ...authorizer.Authorizer) authorizer.Authorizer {
...
 		}
-		return nil
+		if !authorized {
+			continue
+		}
+		return true, "", nil
 	}
 
-	return utilerrors.NewAggregate(errlist)
+	return false, reason, utilerrors.NewAggregate(errlist)
```
Comment (liggitt) [inline pkg/auth/authorizer/union/union.go]: return a joined list of reasons, maybe something like `strings.Join(reasonlist, "\n")`

File: plugin/pkg/auth/authorizer/rbac/rbac.go [hunk 21]
```diff
@@ -62,7 +62,11 @@ func (r *RBACAuthorizer) Authorize(attr authorizer.Attributes) error {
 		}
 	}
 
-	return validation.ConfirmNoEscalation(ctx, r.authorizationRuleResolver, []rbac.PolicyRule{requestedRule})
+	// TODO(nhlfr): Try to find more lightweight way to check attributes than escalation checks.
+	err := validation.ConfirmNoEscalation(ctx, r.authorizationRuleResolver, []rbac.PolicyRule{requestedRule})
+	authorized := bool(err == nil)
+
+	return authorized, "", nil
```
Comment (liggitt) [inline plugin/pkg/auth/authorizer/rbac/rbac.go]: let's keep the error details as the reason

```
if err != nil {
  return false, err.Error(), nil
}
return true, "", nil
```

File: plugin/pkg/auth/authorizer/webhook/webhook.go [hunk 22]
```diff
@@ -180,10 +185,10 @@ func (w *WebhookAuthorizer) Authorize(attr authorizer.Attributes) (err error) {
...
 	if r.Status.Allowed {
-		return nil
+		return true, "", nil
 	}
 	if r.Status.Reason != "" {
-		return errors.New(r.Status.Reason)
+		return false, r.Status.Reason, nil
 	}
-	return errors.New("unauthorized")
+	return false, "", nil
```
Comment (liggitt) [inline plugin/pkg/auth/authorizer/webhook/webhook.go]: I think this whole section can reduce to this (just pass through the Allowed and Reason fields):
```
return r.Status.Allowed, r.Status.Reason, nil
```

File: plugin/pkg/auth/authorizer/webhook/webhook.go [hunk 23]
```diff
@@ -180,10 +185,10 @@ func (w *WebhookAuthorizer) Authorize(attr authorizer.Attributes) (err error) {
 		}
 	}
 	if r.Status.Allowed {
-		return nil
+		return true, "", nil
 	}
 	if r.Status.Reason != "" {
-		return errors.New(r.Status.Reason)
+		return false, r.Status.Reason, nil
```
Comment (liggitt) [inline plugin/pkg/auth/authorizer/webhook/webhook.go]: @ericchiang this looks wrong... if the remote hook returned `{allowed:true,reason:"because we like you"}`, the webhook authorizer would report forbidden... do you agree that just passing through the allowed and reason fields is more correct?

Comment (ericchiang) [inline plugin/pkg/auth/authorizer/webhook/webhook.go]: The current implementation checks `Allowed` before `Reason` so I think your example would correctly return allowed.

But yeah, with the current implementation, you can just `return r.StatusAllowed, r.StatusReason, nil`

Comment (liggitt) [inline plugin/pkg/auth/authorizer/webhook/webhook.go]: ah, right... we'd allow, but lose any reason

File: pkg/apiserver/authz.go (hunk 13 above)
Comment (nhlfr) [inline pkg/apiserver/authz.go]: Yes, you're right. I will remove this and keep that mode only for tests.

File: pkg/apiserver/handlers.go [hunk 24]
```diff
@@ -447,8 +447,13 @@ func WithImpersonation(handler http.Handler, requestContextMapper api.RequestCon
...
 		}
 
-		err := a.Authorize(actingAsAttributes)
+		authorized, reason, err := a.Authorize(actingAsAttributes)
 		if err != nil {
+			internalError(w, req)
+			return
+		}
+		if !authorized {
+			glog.Errorf("Forbidden: %#v, Reason: %s", req.RequestURI, reason)
```
Comment (liggitt) [inline pkg/apiserver/handlers.go]: not sure this should be considered a server error

File: pkg/apiserver/handlers.go [hunk 25]
```diff
@@ -481,12 +486,17 @@ func WithImpersonation(handler http.Handler, requestContextMapper api.RequestCon
...
-		if err == nil {
-			handler.ServeHTTP(w, req)
+		authorized, reason, err := a.Authorize(getAttribs.GetAttribs(req))
+		if err != nil {
+			internalError(w, req)
 			return
 		}
-		forbidden(w, req)
+		if !authorized {
+			glog.Errorf("Forbidden: %#v, Reason: %s", req.RequestURI, reason)
```
Comment (liggitt) [inline pkg/apiserver/handlers.go]: not sure this should be considered a server error

File: plugin/pkg/auth/authorizer/webhook/webhook.go [hunk 26]
```diff
@@ -164,26 +163,26 @@ func (w *WebhookAuthorizer) Authorize(attr authorizer.Attributes) (err error) {
...
 			glog.Errorf("Failed to make webhook authorizer request: %v", err)
-			return err
+			return false, "", err
 		}
 		var statusCode int
-		if result.StatusCode(&statusCode); statusCode < 200 || statusCode >= 300 {
-			return fmt.Errorf("Error contacting webhook: %d", statusCode)
+		result.StatusCode(&statusCode)
+		switch {
+		case statusCode == 401,
```
Comment (liggitt) [inline plugin/pkg/auth/authorizer/webhook/webhook.go]: seems like a 401 or 403 here is still an error... @cjcullen, would you expect a 401/403 from the webhook to mean the user was denied, or that the check itself failed?

File: plugin/pkg/auth/authorizer/webhook/webhook_test.go [hunk 27]
```diff
@@ -357,7 +360,16 @@ func TestTLSConfig(t *testing.T) {
 
 			// Allow all and see if we get an error.
 			service.Allow()
-			err = wh.Authorize(attr)
+			authorized, _, err := wh.Authorize(attr)
+			if tt.wantAuth {
+				if !authorized {
+					t.Errorf("expected succesful authorization")
```
Comment (liggitt) [inline plugin/pkg/auth/authorizer/webhook/webhook_test.go]: successful

File: pkg/apiserver/handlers.go [hunk 28]
```diff
@@ -447,8 +447,13 @@ func WithImpersonation(handler http.Handler, requestContextMapper api.RequestCon
...
 		}
 
-		err := a.Authorize(actingAsAttributes)
+		authorized, reason, err := a.Authorize(actingAsAttributes)
 		if err != nil {
+			internalError(w, req)
+			return
+		}
+		if !authorized {
+			glog.Infof("Forbidden: %#v, Reason: %s", req.RequestURI, reason)
```
Comment (liggitt) [inline pkg/apiserver/handlers.go]: is this debug logging? remove or lower verbosity

File: pkg/apiserver/handlers.go [hunk 29]
```diff
@@ -447,8 +447,13 @@ func WithImpersonation(handler http.Handler, requestContextMapper api.RequestCon
 			actingAsAttributes.Name = name
 		}
 
-		err := a.Authorize(actingAsAttributes)
+		authorized, reason, err := a.Authorize(actingAsAttributes)
 		if err != nil {
+			internalError(w, req)
```
Comment (liggitt) [inline pkg/apiserver/handlers.go]: pass the error to `internalError` and call runtime.HandleError() on it to make sure we log?

File: pkg/apiserver/handlers.go [hunk 30]
```diff
@@ -481,12 +486,17 @@ func WithImpersonation(handler http.Handler, requestContextMapper api.RequestCon
...
-		if err == nil {
-			handler.ServeHTTP(w, req)
+		authorized, reason, err := a.Authorize(getAttribs.GetAttribs(req))
+		if err != nil {
+			internalError(w, req)
 			return
 		}
-		forbidden(w, req)
+		if !authorized {
+			glog.Infof("Forbidden: %#v, Reason: %s", req.RequestURI, reason)
```
Comment (liggitt) [inline pkg/apiserver/handlers.go]: remove or lower verbosity

File: pkg/kubelet/server/server.go [hunk 31]
```diff
@@ -220,8 +220,15 @@ func (s *Server) InstallAuthFilter() {
 		attrs := s.auth.GetRequestAttributes(u, req.Request)
 
 		// Authorize
-		if err := s.auth.Authorize(attrs); err != nil {
-			msg := fmt.Sprintf("Forbidden (user=%s, verb=%s, namespace=%s, resource=%s)", u.GetName(), attrs.GetVerb(), attrs.GetNamespace(), attrs.GetResource())
+		authorized, reason, err := s.auth.Authorize(attrs)
+		if err != nil {
+			msg := fmt.Sprintf("Error (user=%s, verb=%s, namespace=%s, resource=%s)", u.GetName(), attrs.GetVerb(), attrs.GetNamespace(), attrs.GetResource())
+			glog.Errorf(msg)
```
Comment (liggitt) [inline pkg/kubelet/server/server.go]: log the error, even if we don't expose it to the user

File: plugin/pkg/auth/authorizer/webhook/webhook.go (hunk 26 above)
Comment (liggitt) [inline plugin/pkg/auth/authorizer/webhook/webhook.go]: I think this means the webhook is misconfigured and we should return an error, not that the particular user is unauthorized or forbidden

File: plugin/pkg/auth/authorizer/webhook/webhook.go [hunk 32]
```diff
@@ -164,26 +163,26 @@ func (w *WebhookAuthorizer) Authorize(attr authorizer.Attributes) (err error) {
 		if err := result.Error(); err != nil {
```
Comment (liggitt) [inline plugin/pkg/auth/authorizer/webhook/webhook.go]: I actually think the 401 and 403 response codes are getting handled here... otherwise I don't see how the unit test that expects an error in 401/403 cases could be passing. Can you verify that, and remove the 401/403 handling below?

**Comment by liggitt** [comment]
just some logging changes, remove the special 401/403 handling from the webhook authorizer, and I think this will be ready

**Comment by liggitt** [comment]
LGTM, thanks

**Comment by liggitt** [comment]
@k8s-bot please test this issue #28683

File: plugin/pkg/auth/authorizer/webhook/webhook.go (hunk 26 above)
Comment (cjcullen) [inline plugin/pkg/auth/authorizer/webhook/webhook.go]: I'd expect a 401/403 from the webhook to indicate "could not authenticate the request"/"the apiserver is not authorized to ask authorization questions."

Comment (nhlfr) [inline plugin/pkg/auth/authorizer/webhook/webhook.go]: So, should we return 401/403 in such case?

Comment (liggitt) [inline plugin/pkg/auth/authorizer/webhook/webhook.go]: I don't think so. I think a 401/403 from the remote side mean an internal error is presented to the user
